File: lua-ver/decompress.lua
#!/usr/bin/env lua
local ffi = require "ffi"
local vips = require "vips"
local binfmt = require "lua-ver.binfmt"

local args = { ... }
local input_file = args[1] or error("Input file required")
local output_file = args[2] or error("Output file required")

local function yuv_to_rgb_norm(Y, U, V)
  local R = Y + 1.402 * (V - 128)
  local G = Y - 0.344 * (U - 128) - 0.714 * (V - 128)
  local B = Y + 1.772 * (U - 128)
  return R, G, B
end

local function pix_delta_rev(prev, delta, max)
  return (prev + delta) % max
end

local data
if input_file == "-" then
  data = io.stdin:read("*a")
else
  local in_file = io.open(input_file, "rb")
  if not in_file then error("Could not open input file: " .. input_file) end
  data = in_file:read("*a")
  in_file:close()
end

local imgdata = binfmt.buf_to_img(data)
local width, height = imgdata.width, imgdata.height

local pixel_data_ptr = ffi.new("unsigned char[?]", width * height * 3)

print("Decoding...")
local current_x = -8
local current_y = 0

for _, block in ipairs(imgdata.blocks) do
  current_x = current_x + 8
  if current_x >= width then
    current_y = current_y + 8
    current_x = 0
  end

  local corners_orig = {}
  for _, corner in ipairs(block.corners) do
    local oy = (math.floor(corner / 16) / 15) * (block.blockmaxy - block.blockminy) + block.blockminy
    local ou = (math.floor((corner % 16) / 4) / 3) * (block.blockmaxu - block.blockminu) + block.blockminu
    local ov = (math.floor(corner % 4) / 3) * (block.blockmaxv - block.blockminv) + block.blockminv
    table.insert(corners_orig, { oy, ou, ov })
  end

  local function interpolate(tl, tr, bl, br, u, v)
    local top = tl * (1 - u) + tr * u
    local bottom = bl * (1 - u) + br * u
    return top * (1 - v) + bottom * v
  end

  local prevpix = { 0, 0, 0 }
  for blockY = 0, 7 do
    for blockX = 0, 7 do
      local nblock_val                       = block.nblock4bn[blockY + 1][blockX + 1]
      local oy_val                           = math.floor(nblock_val / 16)
      local ou_val                           = math.floor((nblock_val % 16) / 4)
      local ov_val                           = math.floor(nblock_val % 4)

      local dy                               = pix_delta_rev(prevpix[1], oy_val, 16)
      local du                               = pix_delta_rev(prevpix[2], ou_val, 4)
      local dv                               = pix_delta_rev(prevpix[3], ov_val, 4)
      prevpix                                = { dy, du, dv }

      local u_interp, v_interp               = blockX / 7, blockY / 7

      local cy                               = block.interpolatey and
          interpolate(corners_orig[1][1], corners_orig[2][1], corners_orig[3][1], corners_orig[4][1], u_interp, v_interp) or
          (dy / 15) * (block.blockmaxy - block.blockminy) + block.blockminy

      local cu                               = block.interpolateu and
          interpolate(corners_orig[1][2], corners_orig[2][2], corners_orig[3][2], corners_orig[4][2], u_interp, v_interp) or
          (du / 3) * (block.blockmaxu - block.blockminu) + block.blockminu

      local cv                               = block.interpolatev and
          interpolate(corners_orig[1][3], corners_orig[2][3], corners_orig[3][3], corners_orig[4][3], u_interp, v_interp) or
          (dv / 3) * (block.blockmaxv - block.blockminv) + block.blockminv

      local r, g, b                          = yuv_to_rgb_norm(cy, cu, cv)

      local clamped_r                        = math.max(0, math.min(255, math.floor(r + 0.5)))
      local clamped_g                        = math.max(0, math.min(255, math.floor(g + 0.5)))
      local clamped_b                        = math.max(0, math.min(255, math.floor(b + 0.5)))

      local linear_byte_offset               = ((current_y + blockY) * width + (current_x + blockX)) * 3
      pixel_data_ptr[linear_byte_offset]     = clamped_r
      pixel_data_ptr[linear_byte_offset + 1] = clamped_g
      pixel_data_ptr[linear_byte_offset + 2] = clamped_b
    end
  end
end

local out_image = vips.Image.new_from_memory(pixel_data_ptr, width, height, 3, "uchar")

if output_file == "-" then
  local output_buffer = out_image:write_to_buffer(".png")
  io.stdout:write(output_buffer)
else
  out_image:write_to_file(output_file)
end

io.stderr:write("Done.\n")

File: c/decompress.c
#include "binfmt.h"
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <vips/vips.h>

#ifdef _WIN32
#include <winsock2.h>
#else
#include <arpa/inet.h>
#endif

typedef struct {
  double r, g, b;
} RGB_Pixel;

RGB_Pixel yuv_to_rgb_norm(double Y, double U, double V);
int pix_delta_rev(int prev, int delta, int max);
double interpolate(double tl, double tr, double bl, double br, double u,
                   double v);

int main(int argc, char *argv[]) {
  if (VIPS_INIT(argv[0])) {
    vips_error_exit(NULL);
  }

#ifdef _WIN32
  WSADATA wsaData;
  if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
    fprintf(stderr, "WSAStartup failed.\n");
    return 1;
  }
#endif

  if (argc < 3) {
    fprintf(stderr, "Usage: %s <input_file> <output_file>\n", argv[0]);
    return 1;
  }

  const char *input_file = argv[1];
  const char *output_file = argv[2];

  char *data = NULL;
  size_t data_size = 0;

  if (strcmp(input_file, "-") == 0) {
    fprintf(stderr, "Reading from stdin is not supported in this C version.\n");
    return 1;
  } else {
    FILE *in_file = fopen(input_file, "rb");
    if (!in_file) {
      fprintf(stderr, "Could not open input file: %s\n", input_file);
      return 1;
    }
    fseek(in_file, 0, SEEK_END);
    data_size = ftell(in_file);
    fseek(in_file, 0, SEEK_SET);
    data = (char *)malloc(data_size);
    if (!data) {
      fprintf(stderr, "Memory allocation failed for input buffer.\n");
      fclose(in_file);
      return 1;
    }
    fread(data, 1, data_size, in_file);
    fclose(in_file);
  }

  ImgData *imgdata = buf_to_img(data, data_size);
  if (!imgdata) {
    fprintf(stderr, "Failed to decode image data.\n");
    free(data);
    return 1;
  }
  free(data);

  int width = imgdata->width;
  int height = imgdata->height;

  void *pixel_data_ptr;
  size_t pixel_data_size = width * height * 3;
  if (!(pixel_data_ptr = g_malloc(pixel_data_size))) {
    fprintf(stderr, "Failed to allocate memory for pixel data.\n");
    free_imgdata(imgdata);
    return 1;
  }

  fprintf(stderr, "Decoding...\n");
  int current_x = -8;
  int current_y = 0;

  for (int i = 0; i < imgdata->block_count; ++i) {
    BlockData *block = &imgdata->blocks[i];

    current_x += 8;
    if (current_x >= width) {
      current_y += 8;
      current_x = 0;
    }

    double corners_orig[4][3];
    for (int j = 0; j < 4; ++j) {
      uint8_t corner = block->corners[j];
      double oy = (floor(corner / 16.0) / 15.0) *
                      (block->blockmaxy - block->blockminy) +
                  block->blockminy;
      double ou = (floor((corner % 16) / 4.0) / 3.0) *
                      (block->blockmaxu - block->blockminu) +
                  block->blockminu;
      double ov =
          (floor(corner % 4) / 3.0) * (block->blockmaxv - block->blockminv) +
          block->blockminv;
      corners_orig[j][0] = oy;
      corners_orig[j][1] = ou;
      corners_orig[j][2] = ov;
    }

    double prevpix[3] = {0, 0, 0};
    for (int blockY = 0; blockY < 8; ++blockY) {
      for (int blockX = 0; blockX < 8; ++blockX) {
        uint8_t nblock_val = block->nblock4bn[blockY][blockX];
        int oy_val = floor(nblock_val / 16.0);
        int ou_val = floor((nblock_val % 16) / 4.0);
        int ov_val = floor(nblock_val % 4);

        double dy = pix_delta_rev((int)prevpix[0], oy_val, 16);
        double du = pix_delta_rev((int)prevpix[1], ou_val, 4);
        double dv = pix_delta_rev((int)prevpix[2], ov_val, 4);
        prevpix[0] = dy;
        prevpix[1] = du;
        prevpix[2] = dv;

        double u_interp = (double)blockX / 7.0;
        double v_interp = (double)blockY / 7.0;

        double cy, cu, cv;
        if (block->interpolatey) {
          cy = interpolate(corners_orig[0][0], corners_orig[1][0],
                           corners_orig[2][0], corners_orig[3][0], u_interp,
                           v_interp);
        } else {
          cy = (dy / 15.0) * (block->blockmaxy - block->blockminy) +
               block->blockminy;
        }

        if (block->interpolateu) {
          cu = interpolate(corners_orig[0][1], corners_orig[1][1],
                           corners_orig[2][1], corners_orig[3][1], u_interp,
                           v_interp);
        } else {
          cu = (du / 3.0) * (block->blockmaxu - block->blockminu) +
               block->blockminu;
        }

        if (block->interpolatev) {
          cv = interpolate(corners_orig[0][2], corners_orig[1][2],
                           corners_orig[2][2], corners_orig[3][2], u_interp,
                           v_interp);
        } else {
          cv = (dv / 3.0) * (block->blockmaxv - block->blockminv) +
               block->blockminv;
        }

        RGB_Pixel rgb = yuv_to_rgb_norm(cy, cu, cv);

        uint8_t *pixels = (uint8_t *)pixel_data_ptr;
        size_t linear_byte_offset =
            ((current_y + blockY) * width + (current_x + blockX)) * 3;
        pixels[linear_byte_offset] = (uint8_t)fmax(0, fmin(255, round(rgb.r)));
        pixels[linear_byte_offset + 1] =
            (uint8_t)fmax(0, fmin(255, round(rgb.g)));
        pixels[linear_byte_offset + 2] =
            (uint8_t)fmax(0, fmin(255, round(rgb.b)));
      }
    }
  }

  VipsImage *out_image;
  out_image = vips_image_new_from_memory(pixel_data_ptr, pixel_data_size, width,
                                         height, 3, VIPS_FORMAT_UCHAR);
  if (!out_image) {
    vips_error_exit(NULL);
  }
  if (strcmp(output_file, "-") == 0) {

    void *output_buffer;
    size_t output_size;
    if (vips_image_write_to_buffer(out_image, ".png", &output_buffer,
                                   &output_size, NULL) != 0) {
      vips_error_exit(NULL);
    }
    fwrite(output_buffer, 1, output_size, stdout);
    g_free(output_buffer);
  } else {
    if (vips_image_write_to_file(out_image, output_file, NULL) != 0) {
      vips_error_exit(NULL);
    }
  }

  g_free(pixel_data_ptr);
  g_object_unref(out_image);
  free_imgdata(imgdata);
  vips_shutdown();

#ifdef _WIN32
  WSACleanup();
#endif
  fprintf(stderr, "Done.\n");
  return 0;
}

RGB_Pixel yuv_to_rgb_norm(double Y, double U, double V) {
  RGB_Pixel rgb;
  rgb.r = Y + 1.402 * (V - 128);
  rgb.g = Y - 0.344 * (U - 128) - 0.714 * (V - 128);
  rgb.b = Y + 1.772 * (U - 128);
  return rgb;
}

int pix_delta_rev(int prev, int delta, int max) { return (prev + delta) % max; }

double interpolate(double tl, double tr, double bl, double br, double u,
                   double v) {
  double top = tl * (1 - u) + tr * u;
  double bottom = bl * (1 - u) + br * u;
  return top * (1 - v) + bottom * v;
}